pkg/code-generator/schemagen/cue.go

```diff
@@ -0,0 +1,188 @@
+)
+
+// Implementation of JsonSchemaGenerator that uses cuelang (https://github.com/cuelang/cue)
+// packages to translate protobuf definitions into OpenAPI schemas with structural constraints
+// It is modeled after the istio tool: https://github.com/istio/tools/tree/master/cmd/cue-gen
+// TODO (sam-heilbron) This tool has a known flaw that makes it unusable with Gloo Edge protos
+// 	 There are open issues to track this and a GitHub discussion with more context:
+//	 https://github.com/cuelang/cue/discussions/944. We would like to move towards this
+//	 implementation and have added it now to compare schemas generated with this implementation
+//	 to schemas generated with other implementations to ensure consistency.
+type cueGenerator struct {
+	// The Collector used to extract imports for proto files
+	importsCollector collector.Collector
+	absoluteRoot     string
+	protoDir         string
+}
+
+// TODO (sam-heilbron) CUE deprecates OrderedMap
+// 	 We should start using ast.File and other APIs in the ast package to manipulate the schemas
+//	 This requires some more support from ast package thus leaving this as to-do
+
+// Mapping from protobuf message name to OpenApi schema
+type OpenApiSchemas map[string]*openapi.OrderedMap
+
+func NewCueGenerator(importsCollector collector.Collector, absoluteRoot string) *cueGenerator {
+	return &cueGenerator{
+		importsCollector: importsCollector,
+		absoluteRoot:     absoluteRoot,
+		protoDir:         anyvendor.DefaultDepDir,
+	}
+}
+
+func (c *cueGenerator) GetJsonSchemaForProject(project *model.Project) (map[schema.GroupVersionKind]*apiextv1beta1.JSONSchemaProps, error) {
+	protobufExtractor, err := c.getProtobufExtractorForProject(project)
+	if err != nil {
+		return nil, err
+	}
+
+	instances, err := protobufExtractor.Instances()
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert cuelang to openapi
+	openApiGenerator := &openapi.Generator{
+		// k8s structural schemas do not allow $refs, i.e. all references must be expanded
+		ExpandReferences: true,
+	}
+
+	openApiSchemas := OpenApiSchemas{}
+	built := cue.Build(instances)
+	for _, builtInstance := range built {
+		// Avoid generating openapi for irrelevant proto imports.
+		if !strings.HasSuffix(builtInstance.ImportPath, project.ProjectConfig.GoPackage) {
+			continue
+		}
+
+		if builtInstance.Err != nil {
+			return nil, err
+		}
+		if err = builtInstance.Value().Validate(); err != nil {
+			return nil, eris.Errorf("Cue instance validation failed for %s: %+v", project.ProtoPackage, err)
+		}
+		schemas, err := openApiGenerator.Schemas(builtInstance)
+		if err != nil {
+			return nil, eris.Errorf("Cue openapi generation failed for %s: %+v", project.ProtoPackage, err)
+		}
+
+		// Iterate openapi objects to construct mapping from proto message name to openapi schema
+		for _, kv := range schemas.Pairs() {
+			openApiSchemas[kv.Key] = kv.Value.(*openapi.OrderedMap)
+		}
+
+		return c.convertOpenApiSchemasToJsonSchemas(project, openApiSchemas)
```

Comment:
Is there only one builtInstance in built?

Reply:
No I believe each builtinstance corresponds to a protobuf message.

Reply:
Is it returning early here then? It looks like it returns on the first builtInstance?

Reply:
We skip all the irrelevant protos (https://github.com/solo-io/solo-kit/pull/421/files/4831a802ecb7c520a85777b81bef1f4fac22f189#diff-06c6f9e953ed44cf5c7fea458dfd7ebab8492859838751cb5a098d05dd010465R70) so when we come across one, it means that is the top level proto message (ie Gateway, VirtualService, Proxy ...etc)